<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# Параметры подключения к БД очень не желательно хранить в публичном доступе
# После успешного выполнения ноутбука эту ячейку можно удалить, сохранить ноутбук и закоммитить на GitHub
DBNAME ='project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [3]:
# Есть несколько способов избежать отображения предупреждений при выполнении данной работы

# Вариант 1 игнорирование
import warnings
warnings.filterwarnings("ignore")

# Вариант 2 использование sqlalchemy как того просит pandas
from sqlalchemy import create_engine
# Параметры подключения к БД указаны в аргументах к create_engine, если будете использовать этот вариант после запуска ноутбука параметры желательно скрыть
connection = create_engine("postgresql+psycopg2://skillfactory:cCkxxLVrDE8EbvjueeMedPKt@84.201.134.129:5432/project_sql")
connection.connect()
# при этом варианте есть нюанс - везде где в коде встречается зна процентов, его нужно заменить на два знака процентов
# т.е. % -> %% 

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
    select count(*) num_vacancies
    from VACANCIES
'''

In [5]:
# результат запроса
num_vacancies = pd.read_sql_query(query_3_1, connection)
num_vacancies

num_vacancies
0          49197

In [8]:
# Просто как вариант можно рассмотреть отображение с f-strings
print(f'Количество вакансий: {num_vacancies.squeeze()}')

Количество вакансий: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
    select count(*) num_employers
    from EMPLOYERS
'''

In [7]:
# результат запроса
num_employers = pd.read_sql_query(query_3_2, connection)
num_employers

num_employers
0          23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query_3_3 = f'''
    select count(*) num_areas
    from AREAS
'''

In [10]:
# результат запроса
num_areas = pd.read_sql_query(query_3_3, connection)
num_areas

num_areas
0       1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# текст запроса
query_3_4 = f'''
    select count(*) num_industries
    from INDUSTRIES
'''

In [12]:
# результат запроса
num_industries = pd.read_sql_query(query_3_4, connection)
num_industries

num_industries
0             294

***

### выводы по предварительному анализу данных
- Количество вакансий находятся в таблице vacancies  49197
- Количество работодателей в таблица employers  23501
- Количество регионов в таблица areas  1362
- Количество сфер деятельности в таблица industries 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''
    SELECT
        a.name as city,
        count(v.id) as cnt
        
    from areas a
    left join vacancies v on v.area_id = a.id
    GROUP by a.id
    ORDER by cnt desc
    limit 5
'''

In [14]:
# результат запроса
df_2 = pd.read_sql_query(query_4_1, connection)
df_2

city   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса

# Тут хороший пример разных регистров в одном запросе. Это не по PEP 8.
# Если вы выбрали писать в SQL запросы в верхнем регистре, пишите пожалуйста везде в верхнем
# Тут должно быть FROM WHERE, а если пишем в нижнем регистре - select
# Это не ошибка, но код должен быть единообразным.
query_4_2 = f'''
    SELECT
        count(*)  
    from vacancies 
    where salary_from is not null or salary_to is not null

'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
    SELECT
        ROUND(AVG(v.salary_from)) mean_salary
    FROM 
        vacancies v
    WHERE v.salary_from is not Null
    UNION
    SELECT
        ROUND(AVG(v.salary_to))
    FROM 
        vacancies v
    WHERE v.salary_to is not Null
'''


In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

mean_salary
0      71065.0
1     110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
    SELECT
        v.schedule,
        v.employment
    FROM
        vacancies v
    order by 1 desc
    limit 5
'''

In [19]:
# результат запроса
df_4 = pd.read_sql_query(query_4_4, connection)
df_4

schedule           employment
0  Удаленная работа     Полная занятость
1  Удаленная работа     Полная занятость
2  Удаленная работа     Полная занятость
3  Удаленная работа  Частичная занятость
4  Удаленная работа     Полная занятость

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''
    SELECT 
        experience
    FROM
        vacancies
    group by 1 
    order by 1 desc 
    
'''

In [21]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

experience
0       От 3 до 6 лет
1  От 1 года до 3 лет
2           Нет опыта
3         Более 6 лет

***

### Выводы по детальному анализу вакансий
- Мы узнали сколько  вакансий в каждом регионе, Москва	5333 вакансий стала лидером;
- Определили у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой = 24073;
- Опредилили средние значения для нижней и верхней границы зарплатной вилки составило нижнее 71065.0 и верхнее 110537;
- Вывели количество вакансий для каждого сочетания типа рабочего графика и  отсортировали по убыванию количества                  (Удаленная работа	Полная занятость);
- Вывели требуемый опыт работы в порядке возрастания по количеству вакансий, в которых указан данный вариант опыта, самый востребованый опыт От 3 до 6 лет.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
    SELECT
        e.name "Название работодателей",
        count(v.id) as "Количество вакансий"
        
    from EMPLOYERS e
    left join vacancies v on v.employer_id = e.id
     
    GROUP by e.id
    ORDER by 2 desc
    limit 5
'''

In [23]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

Название работодателей  Количество вакансий
0                 Яндекс                 1933
1             Ростелеком                  491
2               Тинькофф                  444
3                   СБЕР                  428
4          Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''
SELECT 
    a.name "Название региона",
    COUNT(distinct e.id) as "Количество работодателей",
    COUNT(distinct v.id) as "Количество вакансий"
FROM   
    AREAS a
LEFT JOIN EMPLOYERS e ON a.id = e.area
LEFT JOIN VACANCIES v ON v.area_id = a.id
WHERE v.id is  null 
GROUP BY "Название региона"
ORDER BY       
     "Количество работодателей" desc,
     "Количество вакансий" desc 
limit 5
'''


In [25]:
# результат запроса
df_2 = pd.read_sql_query(query_5_2, connection)
df_2

Название региона  Количество работодателей  Количество вакансий
0              Россия                       410                    0
1           Казахстан                       207                    0
2  Московская область                        75                    0
3  Краснодарский край                        19                    0
4            Беларусь                        18                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запросаgroup by e.name
query_5_3 = f'''
    select
        e.name "Название работодателей",
        count(distinct v.area_id) "Число регионов"
    from employers e
    join vacancies v on e.id = v.employer_id
    group by e.name
    order by 2 desc
    limit 5
'''      

In [27]:
# результат запроса
df_3 = pd.read_sql_query(query_5_3, connection)
df_3

Название работодателей  Число регионов
0                  Яндекс             181
1              Ростелеком             152
2              Спецремонт             116
3  Поляков Денис Иванович              88
4               ООО ЕФИН               71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''
    select 
        e.name
    from
        employers e
    left join employers_industries e_i on e.id=e_i.employer_id
    group by e.id
    HAVING COUNT(e_i.industry_id) = 0
    ORDER BY 1
    
'''

In [29]:
# результат запроса
df_4 = pd.read_sql_query(query_5_4, connection)
df_4

name
0     1000 мелочей (ИП Гафиатуллина Гульшат Митарисо...
1                                             1001 Soft
2                                             100 печей
3                                              101 Плюс
4                                           1221Системс
...                                                 ...
8414           Ярославский завод упаковочных материалов
8415                           Ярославский филиал ПГУПС
8416    Ярославский ЦГМС - филиал ФГБУ Центральное УГМС
8417                                       Ярпож Казань
8418                            Яшнова Алена Викторовна

[8419 rows x 1 columns]

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''
    select 
        e.name "название компании"
    from
        employers e
    left join employers_industries AS e_i on e.id=e_i.employer_id
    group by e.id
    HAVING COUNT(e_i.industry_id) = 4
    ORDER BY 1
    limit 3
        '''

In [31]:
# результат запроса
df_5 = pd.read_sql_query(query_5_5, connection)
df_5

название компании
0      101 Интернет
1          21vek.by
2              2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
# тут заменил % на %% для проверки работоспособности, далее не менял,
# оставляю на ваше усмотрение какой вариант обработки запросов вам удобнее (с sqlalchemy или без)
query_5_6 = f'''
SELECT
    COUNT(DISTINCT e.id) "количества работодателей"
FROM
    employers e
JOIN
    employers_industries e_i ON e_i.employer_id = e.id
JOIN
    industries i ON i.id = e_i.industry_id
WHERE
    LOWER(i.name) LIKE '%%разработк%%программн%%обеспечен%%';
'''

In [35]:
# результат запроса
df_6 = pd.read_sql_query(query_5_6, connection)
df_6

количества работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
import urllib.parse as req
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) + req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
mln_cities = tuple(pd.read_html(url)[1]['Город'])

In [37]:
# код для получения списка городов-милионников

# Отлично выполнен парсинг
import requests
from bs4 import BeautifulSoup

url='https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response=requests.get(url)
page=BeautifulSoup(response.text,'html.parser')

mln_cities_here=str(page.find('table', class_="standard"))
mln_cities_here=mln_cities_here.split('title="')[3:]
mln_cities=[]
for i in mln_cities_here:
    mln_cities.append(i.split('">')[0])

In [38]:
mln_cities

['Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград']

In [39]:
query_5_7=f'''
SELECT
    a.name AS "регион-миллионников",
    COUNT(*) AS "вакансии компании"
FROM
    vacancies v
JOIN
    employers e ON v.employer_id=e.id

JOIN
    areas a ON a.id=v.area_id
WHERE
    e.name='Яндекс'
GROUP BY a.name
HAVING a.name in {tuple(mln_cities)}

UNION
    
SELECT
    'Total' AS "регион-миллионников",
    SUM("вакансии компании") AS "вакансии компании"
FROM
    (SELECT
        a.name AS "регион-миллионников",
        COUNT(*) AS "вакансии компании"
    FROM
        vacancies v
    JOIN
        employers e ON v.employer_id=e.id
    JOIN
        areas a ON a.id=v.area_id
    WHERE
        e.name='Яндекс'
    GROUP BY a.name
    HAVING a.name in {tuple(mln_cities)}
    ) subquery
ORDER BY
    "вакансии компании" ASC
        
'''

In [40]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

регион-миллионников  вакансии компании
0                 Омск               21.0
1            Челябинск               22.0
2           Красноярск               23.0
3            Волгоград               24.0
4                Пермь               25.0
5               Казань               25.0
6       Ростов-на-Дону               25.0
7                  Уфа               26.0
8               Самара               26.0
9            Краснодар               30.0
10             Воронеж               32.0
11         Новосибирск               35.0
12     Нижний Новгород               36.0
13        Екатеринбург               39.0
14     Санкт-Петербург               42.0
15              Москва               54.0
16               Total              485.0

***

### Выводы по анализу работодателей
- Мы знаем какие работодатели находятся на первом(Яндекс) и пятом(Газпром нефть) месте по количеству вакансий.
- Мы знаем что в Россия в количестве 410 работодателей нет вакансий.
- Мы знаем каждого работодателя и посчитали количество регионов, в которых он публикует свои вакансии:
    - "Яндекс" (181)
	- "Ростелеком" (152)
	- "Спецремонт" (116)
	- "Поляков Денис Иванович" (88)
	- "ООО ЕФИН" (71)
- Мы узнали количество 8419 работодателей, у которых не указана сфера деятельности.
- Мы знаем название компании "2ГИС" что она на третьем местев в алфавитном списке и указано четыре сферы деятельности.
- С помощью запроса «Разработка программного обеспечения» выяснили количества работодателей 3553.
- Для компании «Яндекс» вывели список регионов-миллионников, в которых представлены вакансии компании и количеством вакансий в  них. Также в строке Total -- 485 общее количеством вакансий компании




# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [34]:
# текст запроса
query_6_1 = f'''
    select count(*) "вакансии"
    from vacancies 
    where    ( 
                name ilike '%data%' or
                name ilike '%данн%'
            )
'''

In [35]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

вакансии
0      1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [6]:
# текст запроса
query_6_2 = f'''
select
    count(*) "вакансии начин. дата-сайентистов"
from vacancies 
where  (
            name ilike '%data scientist%' OR 
            name ilike '%data science%' OR
            name ilike '%machine learning%' OR
            name ilike '%машинн%обучен%' OR 
            name ilike '%исследователь данных%' OR 
            name ilike '%ml%' AND NOT name ilike '%html%' 
        )
        AND  (
             name ilike '%junior%' or
             experience ilike '%нет опыта%' or
             employment ilike '%стажировка%'
             )
'''

In [7]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

вакансии начин. дата-сайентистов
0                                51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [4]:
# текст запроса
query_6_3 = f'''
SELECT count(*) "вакансии с навыком SQL или postgres"
FROM vacancies
WHERE   (
            name ilike '%data scientist%' OR 
            name ilike '%data science%' OR
            name ilike '%machine learning%' OR
            name ilike '%машинн%обучен%' OR 
            name ilike '%исследователь данных%' OR 
            name ilike '%ML%' AND NOT name ilike '%HTML%'
        )
        AND ( key_skills ilike '%sql%' or
              key_skills ilike '%postgres%'
        )
'''

In [5]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

вакансии с навыком SQL или postgres
0                                  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [6]:
# текст запроса
query_6_4 = f'''
SELECT count(*) "Python в треб. работодателей к DS"
FROM vacancies
WHERE   (
            name ilike '%data scientist%' OR 
            name ilike '%data science%' OR
            name ilike '%machine learning%' OR
            name ilike '%машинн%обучен%' OR 
            name ilike '%исследователь данных%' OR 
            name ilike '%ML%' AND NOT name ilike '%HTML%'
        )
        AND ( key_skills ilike '%python%'
        )
'''

In [7]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

Python в треб. работодателей к DS
0                                357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [28]:
query_6_5 = f''' 
SELECT 
    round(avg(LENGTH(key_skills)-LENGTH(replace(key_skills, CHR(9), '')) +1), 2) as "навыков в среднем"
FROM vacancies 
WHERE (
            name ilike '%data scientist%' OR 
            name ilike '%data science%' OR
            name ilike '%machine learning%' OR
            name ilike '%машинн%обучен%' OR 
            name ilike '%исследователь данных%' OR 
            name like '%ML%' AND name  NOT like '%HTML%'
      )
            AND key_skills IS NOT NULL        
               
     
'''

In [29]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

навыков в среднем
0               6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [35]:
# текст запроса
query_6_6 = f'''        
    select 
            experience "требуемый опыт",
        ROUND(avg(COALESCE((salary_to + salary_from)/2, salary_from, salary_to)), 0) AS "средняя зарплата"
            
    from vacancies 
        
    where   
        (name ILIKE '%data scientist%' OR
         name ILIKE '%data science%' OR 
         name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
        )
    AND (salary_from is not null OR salary_to is not null)

    GROUP BY experience
 
'''

In [36]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

требуемый опыт  средняя зарплата
0           Нет опыта           74643.0
1  От 1 года до 3 лет          139675.0
2       От 3 до 6 лет          243115.0

***

### Выводы по предметному анализу
- Количество вакансий имеет отношение к данным со слова 'data' или 'данн' будет 1771.
- Для начинающего дата-сайентиста есть 51 вакансия.
- Для DS в которых в качестве ключевого навыка указан SQL или postgres есть 229 вакансий.
- Для DS в которых в качестве ключевого навыка указан Python есть 357 вакансий.
- Ключевых навыков в среднем указывают в вакансиях для DS 6.44.
- Выяснили, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет 243115.00.

In [18]:
connection.close()

# Общий вывод по проекту

#### подведем итог исследования, обобщите выводы
- В этом проекте наглядно видно с чем сталкивается Data Science в работе с SQL таблицами и как можно маневрировать получая нужный результат создавая новые таблицы для выводов и анализа в данном случи вакансий на рынке труда. Благодарю за этот кейс))

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

***
Тут очень хотелось увидеть дополнительные исследования данных
***